File: src/lib/rpc.ts
import { invoke } from "@tauri-apps/api/core";
import { getDownloadStats } from "$lib/stores/download-stats.svelte";

export type RpcSource = "focus" | "music" | "video" | "course" | "reading";

export type RpcSetSourceArgs = {
  source: RpcSource;
  details: string;
  state: string;
  duration?: number;
  position?: number;
  paused?: boolean;
  largeImageKey?: string | null;
};

export async function rpcSetSource(args: RpcSetSourceArgs): Promise<void> {
  try {
    await invoke("rpc_set_source", {
      source: args.source,
      details: args.details,
      state: args.state,
      duration: args.duration ?? 0,
      position: args.position ?? 0,
      paused: args.paused ?? false,
      largeImageKey: args.largeImageKey ?? null,
    });
  } catch {}
}

export async function rpcClearSource(source: RpcSource): Promise<void> {
  try {
    await invoke("rpc_clear_source", { source });
  } catch {}
}

export async function rpcSyncIdleStats(): Promise<void> {
  const stats = getDownloadStats();
  try {
    await invoke("rpc_set_idle_stats", {
      downloadsCount: stats.totalDownloads,
      totalBytes: stats.totalBytes,
    });
  } catch {}
}

export async function rpcTestConnection(): Promise<{ ok: boolean; reason?: string }> {
  return await invoke("rpc_test_connection");
}

File: src-tauri/src/commands/rpc.rs
use serde_json::Value;

use crate::core::rpc;
use crate::storage::config;

#[tauri::command]
pub async fn rpc_test_connection(app: tauri::AppHandle) -> Result<Value, String> {
    let settings = config::load_settings(&app);
    rpc::test_connection(settings.rpc.clone()).await
}

#[tauri::command]
pub async fn rpc_set_source(
    app: tauri::AppHandle,
    source: String,
    details: String,
    state: String,
    duration: i64,
    position: i64,
    paused: bool,
    large_image_key: Option<String>,
) -> Result<Value, String> {
    let settings = config::load_settings(&app);
    rpc::set_presence_source(
        settings.rpc.clone(),
        source,
        details,
        state,
        duration,
        position,
        paused,
        large_image_key,
    )
    .await
}

#[tauri::command]
pub async fn rpc_clear_source(app: tauri::AppHandle, source: String) -> Result<Value, String> {
    let settings = config::load_settings(&app);
    rpc::clear_source(settings.rpc.clone(), source).await
}

#[tauri::command]
pub async fn rpc_set_idle_stats(
    app: tauri::AppHandle,
    downloads_count: u64,
    total_bytes: u64,
) -> Result<Value, String> {
    let settings = config::load_settings(&app);
    rpc::set_idle_stats(settings.rpc.clone(), downloads_count, total_bytes).await
}
